### PR TITLE
Remove caveat of not supporting different directions of ordering

### PR DESCRIPTION
**Pull request by yeahframeoff**

### CODE REVIEW

File: cursor_pagination.py [hunk 1]
```diff
@@ -104,12 +80,39 @@ def page(self, first=None, last=None, after=None, before=None):
...
         position = self.decode_cursor(cursor)
 
-        is_reversed = self.ordering[0].startswith('-')
-        queryset = queryset.annotate(_cursor=Tuple(*[o.lstrip('-') for o in self.ordering]))
-        current_position = [Value(p, output_field=TextField()) for p in position]
-        if reverse != is_reversed:
-            return queryset.filter(_cursor__lt=Tuple(*current_position))
-        return queryset.filter(_cursor__gt=Tuple(*current_position))
+        filtering = Q()
+        q_acc = []
```
Comment (Drarok) [inline cursor_pagination.py]: What is `q_acc` short for? Can we give it a descriptive name?

File: cursor_pagination.py [hunk 2]
```diff
@@ -104,12 +80,39 @@ def page(self, first=None, last=None, after=None, before=None):
...
+
+            q = {key_cmp: value}
+            for q_eq in q_acc:
+                q.update(q_eq)
+
+            filtering |= Q(**q)
+
+            q_acc.append({key_eq: value})
+
+        return queryset.filter(filtering)
```
Comment (Drarok) [inline cursor_pagination.py:142]: This whole block feels like it has too much whitespace – take a look at the other methods for comparison.

File: cursor_pagination.py [hunk 3]
```diff
@@ -104,12 +80,39 @@ def page(self, first=None, last=None, after=None, before=None):
...
+            else:
+                key_cmp = "{}__gt".format(o)
+
+            key_eq = "{}__exact".format(o)
+
+            value = Value(position, output_field=TextField())
+
+            q = {key_cmp: value}
+            for q_eq in q_acc:
+                q.update(q_eq)
```
Comment (Drarok) [inline cursor_pagination.py]: Again, the variable names here make this extremely hard to follow.

File: cursor_pagination.py [hunk 4]
```diff
@@ -104,12 +80,39 @@ def page(self, first=None, last=None, after=None, before=None):
...
+
+            is_reversed = ordering.startswith('-')
+            o = ordering.lstrip('-')
+
+            if reverse != is_reversed:
+                key_cmp = "{}__lt".format(o)
+            else:
+                key_cmp = "{}__gt".format(o)
+
+            key_eq = "{}__exact".format(o)
```
Comment (Drarok) [inline cursor_pagination.py]: `key_cmp`? `key_eq`?

**Review by Drarok — CHANGES_REQUESTED** [review]

**Comment by yeahframeoff** [comment]
@Drarok what tests does it need so that it could be merged?

File: cursor_pagination.py [hunk 5]
```diff
@@ -104,12 +80,67 @@ def page(self, first=None, last=None, after=None, before=None):
...
+
+        # Suppose we have ordering ("field1", "-field2", "field3")
+        # (note negation 2nd item),
+        # and corresponding cursor values are ("value1", "value2", "value3"),
+        # `reverse` is False.
+        # In order to apply cursor, we need to generate a following WHERE-clause:
+
+        # WHERE ((field1 < value1) OR
+        #     (field1 = value1 AND field2 > value2) OR
+        #     (field1 = value1 AND field2 = value2 OR field3 < value3).
```
Comment (cmmartti) [inline cursor_pagination.py]: Typo? It looks like the "OR" should be "AND".

**Comment by mjtamlyn** [comment]
Hi! This looks like it has all the correct logic in place. A few thoughts:

1. Is there any performance implication in the database of the longer query structure? If so, could we drop to the tuple style code when the ordering is simpler. Some real-ish benchmarks would be useful.
2. Can we add some more tests of some more complex situations - 3 or 4 variables with some different orderings (i.e. `a, -b, -c, d`, `a, b, -c, -d` - perhaps even all permutations). Won't exercise any new code but will encourage the feeling that this works.
3. The above OR/AND comment needs updating

This should then allow us to update and merge it.

**Comment by nitely** [comment]
> Is there any performance implication in the database of the longer query structure? If so, could we drop to the tuple style code when the ordering is simpler. Some real-ish benchmarks would be useful.

You can expect this query to be slower than the offset/limit approach, as it won't use the index. There is a variation that will use the index for the first field only, so it'll still be slow in most cases.

**Comment by AndrewIngram** [comment]
Just bumping this because we found that there's limited utility for this library without support for mixed sort directions. We ended up vendoring this branch for internal use. This PR essentially implements the approach i've had good success with for several years.

**Comment by patrick91** [comment]
> Just bumping this because we found that there's limited utility for this library without support for mixed sort directions. We ended up vendoring this branch for internal use. This PR essentially implements the approach i've had good success with for several years.

Thanks Andrew! I'll make a not to merge this and make a new major release in the next few weeks!

**Comment by patrick91** [comment]
I'll merge this and make a release soon! sorry it took a while to get this in 😊
